### PR TITLE
feat(recipe): add MBPP code generation recipe with GRPO   

### PR DESCRIPTION
**Pull request by dhruvnigam93**
Adds the first code generation recipe to verl-recipe. Uses GRPO with execution-based rewards on MBPP (sanitized) to train Qwen2.5-1.5B-Instruct for python code generation.Addresses verl-project/verl#5531.                                                                                
                                                                                                                                                                                                                                                   
## Why this recipe?                                                                                                                                                                                                                                    
                  
verl-recipe has extensive math/reasoning recipes (r1, dapo, prime, open_math_reasoning) but no code generation recipe. Similar to math code tasks also often have objective, deterministically verifiable rewards through test cases. 
          
This recipe is intentionally minimal and self-contained: one model, one dataset, one GPU, ~45 minutes to train. The goal is to provide a working reference for training code generation with verifiable rewards using GRPO. Extensions to larger models (7B+ with LoRA), bigger datasets (APPS, LiveCodeBench), and container-based sandboxing (SandboxFusion) can build on this foundation.
                                                                                                                                                                                                                                                   
## Recipe Design       

  - *Model*: Qwen2.5-1.5B-Instruct (full-weight, single A100 40GB)                                                                                                                                                                               
  - *Dataset*: MBPP-sanitized — 120 train / 257 test problems with assert-based test cases. We chose this specifically because of the availability of clean, easily executable test cases.
  - *Reward*: fraction of test cases passed, evaluated via subprocess sandboxing with 10s timeout                                                                                                                                                                                                                                                                                        
  - *Sandboxing*: stdlib-only subprocess with `multiprocessing.spawn` — avoids importing verl/torch in child processes, shadows `input()`, kills on timeout  
  
## Results                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                   
  25 epochs of GRPO on a single A100 40GB (~45 min):                                                                                                                                                                                               
  
  | Stage | MBPP Test Pass Rate |                                                                                                                                                                                                                  
  |-------|---------------------|
  | Qwen2.5-1.5B-Instruct (baseline) | 61.6% |                                                                                                                                                                                                     
  | After GRPO | **69.7%** |           
<img width="808" height="383" alt="Screenshot 2026-04-01 at 3 11 13 PM" src="https://github.com/user-attachments/assets/0a02c09c-2859-4126-baaf-d6d470464e55" />

### CODE REVIEW

File: mbpp_code_gen_grpo/create_mbpp_dataset.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    rl_hint_tests = test_list[:n_hint] if n_hint > 0 else None
+    rl_user_content = compose_user_text(
+        mbpp_prompt,
+        global_suffix=global_suffix,
+        tests_to_show=rl_hint_tests,
+        tests_label="Your implementation should pass tests like the following (more tests will be used for evaluation):",
+    )
+
+    extra_info: dict[str, Any] = {
+        "test_list": test_list,
```
[the remark this answers is not in the document]
Comment (dhruvnigam93) [reply mbpp_code_gen_grpo/create_mbpp_dataset.py:105]: Fixed the inconsistency in 
https://github.com/verl-project/verl-recipe/pull/78/commits/2b335428b897e804c9b908545659655ddbb9aac8

File: mbpp_code_gen_grpo/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+Write your complete solution in a single markdown fenced code block using the language tag `python (open with ```python on its own line, close with `).
+```
+
+The code for generating the entire verl-compatible RL parquet files for training and validation is in the `create_mbpp_dataset.py` script.
+
+## Reward Function
+
+ The reward function is a verl-compatible reward function that takes the generated code and the test cases as input and returns the reward. The reward is the fraction of test cases that pass using the generated solution. If all the test cases pass, the reward is 1.0. If none of the test cases pass, for any reason, the reward is 0.0. The reward function is defined in the `reward_function.py` script.
+
+To make sure that the generated code does not compromise the training process or the entire system, we use a sandboxed environment to execute the model generated code. The sandboxed environment is created using the `mbpp_exec_isolated.py` script using an independent process and a timeout of 10 seconds. If the code takes longer than 10 seconds to execute, the process is terminated and the reward is 0.0.
```
[the remark this answers is not in the document]
Comment (dhruvnigam93) [reply mbpp_code_gen_grpo/README.md:76]: Agreed. My thinking was that process isolation with timeouts should be a good enough starting point for this recipe. 

But if a robust sandbox is a strict requirement, I will incorporate it.

[the remark this answers is not in the document]
Comment (dhruvnigam93) [reply mbpp_code_gen_grpo/README.md:76]: Agreed. My thinking was that process isolation with timeouts should be a good enough starting point for this recipe.

But if a robust sandbox is a strict requirement, I will incorporate it.

**Comment by dhruvnigamd11** [comment]
@wuxibin89 Could you take a look at this when you get a chance? This implements the MBPP code generation recipe from the RFC at verl-project/verl#5531. Happy to address any feedback!

**Comment by dhruvnigam93** [comment]
@wuxibin89 hey, just putting this up for the last time. let me know if this PR is useful.
